File: elixir/lib/day3.ex
defmodule Aoc.Day3 do
  def run_part1 do
    IO.inspect count_trees(3, 1)
  end

  def run_part2 do
    [{1, 1}, {3, 1}, {5, 1}, {7, 1}, {1, 2}]
    |> Enum.map(fn {right, down} -> count_trees(right, down) end)
    |> Enum.reduce(&(&1 * &2))
    |> IO.inspect
  end

  defp count_trees(right, down) do
    "day3"
    |> Aoc.Input.stream_lines_characters()
    |> Stream.take_every(down)
    |> Stream.zip(Stream.iterate(0, &(&1 + right)))
    |> Enum.count(fn {row, col} ->
      Enum.at(row, rem(col, length(row))) == "#"
    end)
  end
end

File: 2020/rust/src/day3.rs
use std::iter;

use super::utils;

pub fn run_part1() -> usize {
  let run_part2 = count_trees(3, 1);

  println!("{}", run_part2);

  run_part2
}

pub fn run_part2() -> usize {
  let result: usize = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    .iter()
    .map(|(right, down)| count_trees(*right, *down))
    .product();

  println!("{}", result);

  result
}

fn count_trees(right: usize, down: usize) -> usize {
  let mut curr = 0;
  
  let cols = iter::repeat_with(|| {
    let tmp = curr;
    curr += right;
    tmp
  });

  let result = utils::file::read_lines("../_inputs/day3.txt")
    .step_by(down)
    .zip(cols)
    .filter(|(row, col)| {
      let chars: Vec<char> = row.chars().collect();

      return chars[col % chars.len()] == '#';
    })
    .count();

  return result;
}
